Improve -Dy debugging

khwilliamson opened the pull request:
Commit 6ceb4087860c6ef8e86e0c252feb738d635e9e3f added a way to cleanly output UTF-8 tr/// values.  This commit uses that to improve the debug output of compiling and running tr///.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
Create embed.fnc entry for pv_display_flags_

This creates an ARGS_ASSERT for this function.  Previously, the code was
using the one for plain pv_display(), which is kind of ugly.  Now there
is a macro for each function

khwilliamson committed:
Convert _pv_display_flags to legal name

It is undefined behavior in C to have a file-level symbol begin with an
underscore.  This is a static function; there is no need for it to have
a leading or trailing underscore to mark it as internal

jkeenan commented:
> Commit [6ceb408](https://github.com/Perl/perl5/commit/6ceb4087860c6ef8e86e0c252feb738d635e9e3f) added a way to cleanly output UTF-8 tr/// values. This commit uses that to improve the debug output of compiling and running tr///.
> 
>     * This set of changes does not require a perldelta entry.

Could you provide a before/example of the effect this p.r. is intended to have?

khwilliamson committed:
op.c: Combine some adjacent -Dy source lines

The same printf can handle two things.

khwilliamson committed:
Add detail to -Dy debugging

Commit 6ceb4087860c6ef8e86e0c252feb738d635e9e3f added a way to cleanly
output UTF-8 tr/// values.  This commit uses that to improve the debug
output of compiling and running tr///.

For a simple tr of of transliterating Greek capital letters to
lowercase, the output of 'perl -Dy' has these added lines:

 > op.c: 6553: Compiling tr/*t/*r/; /c=0; /d=0; /s=0
 > *t is '\x{391}-\x{3a9}'
 > *r is '\x{3b1}-\x{3c9}'

Before the aforementioned commit the minus sign indicating a range would
not have rendered properly; so things like that were omitted from the
debug output.

The output also now includes special mention of the special casing where
the input is complemented, and/or some characters not being translated
or get deleted.

khwilliamson committed:
op.c: Change a -Dy line to yv; and another vice versa

While debugging this code, I found that the final generated map was
unneeded and too large for most instances; so change that to not be
output without the verbose option.

Conversely, I found that I needed to often see the state of things after
the first pass, so change to make that not verbose.

bulk88 commented on op.c:
```diff
@@ -6528,6 +6547,13 @@ S_pmtrans(pTHX_ OP *o, OP *expr, OP *repl)
+    DEBUG_y(PerlIO_printf(Perl_debug_log,
+                          "%s: %d: Compiling tr/*t/*r/; /c=%d; /d=%d; /s=%d\n"
+                          "*t is '%s'\n*r is '%s'\n",
+                          __FILE__, __LINE__, complement, del, squash,
```
__FILE__ is a CPP foldable "" literal on all OSes, skip the "%s" unless the C string literal in this debug fmt string has reuse somewhere else in libperl the binary.

mauke replied:
1. You're trying to micro-optimize debugging output. Don't.
2. Interpolating `__FILE__` in the format string is not just harder to read, it also breaks if `__FILE__` contains a `%`. (In this case it doesn't, but why make the reader worry?)

bulk88 commented on op.c:
```diff
@@ -6904,9 +6935,25 @@ S_pmtrans(pTHX_ OP *o, OP *expr, OP *repl)
+                        PerlIO_printf(Perl_debug_log, "TR_SPECIAL_HANDLING\n");
+                    }
+                    else {
+                        PerlIO_printf(Perl_debug_log,
```
Dont execute function Perl_debug_log multiple times in a row.

mauke replied:
Do you mean `PerlIO_stderr()`? Why not? (If your answer is performance, you're wrong.)

bulk88 commented on op.c:
```diff
@@ -7431,8 +7488,32 @@ S_pmtrans(pTHX_ OP *o, OP *expr, OP *repl)
+            else if ((UV) tbl->map[tbl->size] == TR_SPECIAL_HANDLING) {
+                if (! del) {
+                    const int size = tbl->size;
+                    Perl_croak(aTHX_ "panic: Unexpected value %x in [%d]",
```
Use nocontext variant that doesn't require pushing a my_perl ptr from a c auto to a volatile outgoing C  argument. Less instructions on a panic branch tat will never execute.

mauke replied:
Now you're trying to nano-optimize debugging code. Why?

mauke commented on op.c:
```diff
@@ -7431,8 +7488,32 @@ S_pmtrans(pTHX_ OP *o, OP *expr, OP *repl)
+            else if ((UV) tbl->map[tbl->size] == TR_SPECIAL_HANDLING) {
+                if (! del) {
+                    const int size = tbl->size;
+                    Perl_croak(aTHX_ "panic: Unexpected value %x in [%d]",
```
This could be
```suggestion
                    croak("panic: Unexpected value %x in [%d]",
```

khwilliamson replied:
Thanks; this commit predates that ability.
